calc/man/calc.1
.\" Process this file with
.\" groff -man -Tascii calc.1
.\"
.TH CALC 1 "APRIL 2007" Linux "User Commands"
.SH NAME
calc \- console calculator
.SH SYNOPSIS
.B calc 
[options] "expression"

.B calc
-m script "path"
.SH DESCRIPTION
.B calc
calculates the given input term and returns the result as text.
Please use \" \" at the given expression because otherwise you
will have problems with the shell interpretor. If you give 
.B calc
more than one expression, all expression will be put
together.
.SH CALCULATION SYNTAX
.IP "standard calculations"
2*(3+4/9)^3
.IP "standard operations"
+ - * / ^ %
sin cos tan asin acos atan sinh cosh tanh asinh acosh atanh
ln log
& | ! && || ~ >> << x
rnd sqrt curt root integ d/dx
.IP "sqare root and cubic root"
sqrt12
curt8
.IP "trigonometric functions and logarithms"
sin3.64+ln5
.IP "other roots"
fourth root of nine: 4root9 
.IP "differential calculation"
differential calculation of 2x^2-3x+3 at x=9: d/dx(2x^2-3x+3,9)
.IP integration
integration of sinx+cosx between -1 and 3.2: integ(sinx+cosx,-1,3.2)
.IP "set variables"
save the result of a calculation in variable A: 12*45/2->A
.IP "recall variables"
multiply the saved value with 2: A*2
.IP "use last result"
The last result is saved in the answer memory: ans+1
.SH CONSTANTS
.IP e
euler constant
.IP pi
pi; example: sin(pi/2)
.SH NUMBERS
.IP standard
12.34
.IP "with exponent"
12.23e-5
.IP complex
12i-5

.SH OPTIONS
.IP "-o, --output [output length]"
The argument is a number between 2 and the maximal possible output length (normally 18).
.IP "-m, --mode [calculator mode]"
The mode can be std (default, normal scientific calculatons), base (for logic functions and base-n calculatons) or script (for executing scripts). In scripting mode, the last argument must be the path to the script file.
.IP "-c, --complex"
This option enables complex numbers for the calculation. They are switched off by default.
.IP "-b, --base [calculator base]"
The argument must be one of hex (hexadecimal), dec (decimal, default), oct (octal) and bin (binary). This option can only be used when -m base is set.
.IP "-a, --angle [angle type]"
The argument must be one of deg (degrees), rad (radiant), gra (grade).
.IP "-v, --version"
Output the version of Calc.
.IP "-h, --help"
Outputs a short help.
.SH SCRIPTING
The script interpreter of calc uses a C-like programming syntax.
The following command are supported.
.IP "if expression"

if(condition)

  command for condition true;

else command for condition false;

.IP "while loop"

while(condition)

  command that is run while condition is true;

.IP "for loop"

for(initialisation; condition; count-command)
  command that is run while condition is true;

.IP "print command"

print( output to print out );

.IP "getline command"

getline; // read a text line from stdin and returns this text

.IP "getkey command"

getkey; // return the key pressed by the user
        // blocks until the user presses any key

.IP "keystate command"

keystate; // returns the key pressed by the user or 0 if no key was pressed
          // nonblocking

.IP "sleep command"

sleep( time in microseconds ); //Script is stops for the given time

.IP "compare operators"

==, != equal, unequal

>=, <=, >, < greather or equal, less or equal, greather, less

.IP "set operators"

-> calculator set operator

= C-like set operator

.IP "type converting"

(float) floating point value

(int) integer value

(string) text value

(bool) boolean value

  


.SH BUGS
No bugs known!
.IP "If you find one, please report it by e-mail to the author."
.SH AUTHOR
Rainer Strobel <rainer1223@users.sourceforge.net>
.SH "SEE ALSO"
http://extcalc-linux.sourceforge.net


calc/ChangeLog
alpha:
- first revision
- modified extcalc math algorithms to work in calc
- modified accerelated algorithms for integration and differentiation for calc

0.2.2:
- inserted new extcalc algorithms
- enabled command-line options -v (version) -h (help) -o (output length) -a (angle)

0.3.0:
- updated to extcalc 0.3.0 algorithms

0.3.1:
- variables can be used
- answer memory works

0.4.0
- problems with roots solved
- update to calculation algorithms v0.4.0

0.4.5
- supports scripting
- updated manpage

0.5.0
- full scripting support (including subprograms, arrays, brake, continue, ...)
- script debug messages
- documentation update
- bugfixed calculator algorithms

0.6.0
- updated to Extcalc calculator algorithms v0.6.0
- fixed bug in script interpreter (BSD)

0.6.5
- updated to Execalc algorithms v0.6.5
- support for complex numbers

0.7.0
- updated to Extcalc algorithms v0.7.0
- added option to switch complex numbers on and off
- added preferences preprocessor
- fixed bug in installer


calc/src/calc.cpp
#include <stdio.h>
#include "global.h"
#include <sys/time.h>
#include <float.h>
#include <stdlib.h>
#include "consolecalc.h"

///////////////////////////////////////Calc///////////////////////////////////////
//  Calc is a scientific calculator for the text console.                       //
//  It is a sub-project of Extcalc and it uses the calculaton algorithms of     //
//  Extcalc.                                                                    //
//  Calc is very simple to use. Just start the program with your calculaton     //
//  and it will return the result.                                              //
//                                                                              //
//////////////////////////////////////////////////////////////////////////////////


///////////////////////////////////////TODO///////////////////////////////////////
//                                                                              //
//  - write all needed subroutines for calculator                           ok  //
//  - make all needed changes that the calc algorithms run standalone       ok  //
//  - port accerelated calcualtion algorithm form extcalc to calc           ok  //
//  - make all algorithms compatible with extcalc                           ok  //
//  - enable command line parameters for setting output format              ok  //
//  - enable command line parameters for different angle formats            ok  //
//  - enable base-n calculatons                                             ok  //
//  - enable logic functions                                                ok  //
//  - make usage of variables possible                                      ok  //
//  - write documentation and manpage                                           //
//                                                                              //
//////////////////////////////////////////////////////////////////////////////////


///////////////////////////////////////Bugs///////////////////////////////////////
//
//	- Base mode doesn't work												fixed
//
//
//
//////////////////////////////////////////////////////////////////////////////////




int main(int argc,char**argv)
{
	Variable *vars=new Variable[VARNUM];
	Number value;

	for(int c=0; c<VARNUM;c++)
		vars[c]=0.0;
	Preferences *pref=new Preferences;
	pref->angle=RAD;
	pref->calcType=SCIENTIFIC;
	pref->base=DEC;
	pref->complex=false;
	bool ask=true;
	char*input=NULL;
	bool script=false;
	Script*s;
	ThreadSync scriptData;
	scriptData.vars=new Number*[VARNUM];
	for(int c=0; c<VARNUM;c++)
	{
		scriptData.vars[c]=(Number*)malloc(sizeof(Number));
		scriptData.numlen[c]=1;
		for(int c1=0; c1<VARDIMENSIONS; c1++)
			scriptData.dimension[c][c1]=1;
		scriptData.vars[c][0].type=NNONE;
		scriptData.vars[c][0].cval=NULL;
	}

	
#ifndef NO_LONG_DOUBLE
	pref->outputLength=pref->precision=LDBL_DIG;
	int maxLength=LDBL_DIG;
#else 
	pref->outputLength=pref->precision=DBL_DIG;
	int maxLength=DBL_DIG;
#endif

	char*homedir=getenv("HOME");
	char*path=new char[strlen(homedir)+17];
	memcpy(path,homedir,strlen(homedir));
	memcpy(&path[strlen(homedir)],"/.calcvariables",16);

	struct timeval rndTime;
	gettimeofday(&rndTime,NULL);
	srand(rndTime.tv_usec*rndTime.tv_sec);
	
	int fileLength;
	struct stat fileStat;
	char*variablesString;

	char*var;

	if(lstat(path,&fileStat) == 0)
	{
		FILE*variablesFile=fopen(path,"r");
		if(variablesFile!=NULL)
		{
			fileLength=fileStat.st_size;
			variablesString=new char[fileLength+1];
			fread(variablesString,fileLength,1,variablesFile);
			
			int varStart=0;
			int varEnd=0;
			int varCount=0;
			bool real=true;
		
		
			for(int c=0; c<fileLength && varCount<27; c++)
			{
				if(variablesString[c]=='\n' || c == fileLength-1)
				{
					varEnd=c;
					var=new char[varEnd-varStart+1];
					memcpy(var,&variablesString[varStart],varEnd-varStart);
					var[varEnd-varStart]=(char)0;
					scriptData.vars[varCount][0].type=NFLOAT;
					if(real)
					{
#ifndef NO_LONG_DOUBLE
					vars[varCount]=strtold(var,NULL);
#else 
					vars[varCount]=strtod(var,NULL);
#endif
					real=false;
					}
					else 
					{
#ifndef NO_LONG_DOUBLE
					scriptData.vars[varCount][0].fval=Complex(vars[varCount],strtold(var,NULL));
#else 
					scriptData.vars[varCount][0].fval=Complex(vars[varCount],strtod(var,NULL));
#endif
						varCount++;
						real=true;
					}
					varStart=varEnd+1;
					delete[] var;
				}
			}
		fclose(variablesFile);
		}
	}


	for(int c=1; c<argc; c++)
	{
		if(strcmp(argv[c],"-a")==0 || strcmp(argv[c],"--angle")==0)
		{
			if(argc>c)
			{
				if(strncmp(argv[c+1],"deg",3)==0)
					pref->angle=DEG;
				else if(strncmp(argv[c+1],"rad",3)==0)
					pref->angle=RAD;
				else if(strncmp(argv[c+1],"gra",3)==0)
					pref->angle=GRA;
				else {
					printf("Invalid argument for -a!\n\nPossible Arguments:\n-a deg\n-a rad\n-a gra\n\nPlease use man calc for further information\n");
					delete pref;
					delete[] vars;
					return 0;
				}
				c++;
			}
			else {
				printf("Invalid argument for -a!\n\nPossible Arguments:\n-a deg\n-a rad\n-a gra\n\nPlease use man calc for further information\n");
				delete pref;
				delete[] vars;
				return 0;
			}
		}
		else if(strcmp(argv[c],"-m")==0 || strcmp(argv[c],"--mode")==0)
		{
			if(argc>c)
			{
				if(strncmp(argv[c+1],"base",3)==0)
					pref->calcType=BASE;
				else if(strncmp(argv[c+1],"std",3)==0)
					pref->calcType=SCIENTIFIC;
				else if(strncmp(argv[c+1],"script",3)==0)
				{
					pref->calcType=SCIENTIFIC;
					script=true;
				}
				else {
					printf("Invalid argument for -m!\n\nPossible Arguments:\n-m base\n-m std\n\nPlease use man calc for further information\n");
					delete pref;
					delete[] vars;
					return 0;
				}
				c++;
			}
			else {
				printf("Invalid argument for -m!\n\nPossible Arguments:\n-m base\n-m std\n\nPlease use man calc for further information\n");
				delete pref;
				delete[] vars;
				return 0;
			}
		}
		else if(strcmp(argv[c],"-b")==0 || strcmp(argv[c],"--base")==0)
		{
			if(argc>c)
			{
				if(strncmp(argv[c+1],"bin",3)==0)
					pref->base=BIN;
				else if(strncmp(argv[c+1],"hex",3)==0)
					pref->base=HEX;
				else if(strncmp(argv[c+1],"oct",3)==0)
					pref->base=OCT;
				else if(strncmp(argv[c+1],"dec",3)==0)
					pref->base=DEC;
				else {
					printf("Invalid argument for -m!\n\nPossible Arguments:\n-m base\n-m std\n\nPlease use man calc for further information\n");
					delete pref;
					delete[] vars;
					return 0;
				}
				c++;
			}
			else {
				printf("Invalid argument for -b!\n\nPossible Arguments:\n-b dec\n-b hex\n-b bin\n-b oct\n\nPlease use man calc for further information\n");
				delete pref;
				delete[] vars;
				return 0;
			}
		}
		else if(strcmp(argv[c],"-o")==0 || strcmp(argv[c],"--output")==0)
		{
			if(argc>c)
			{
				int outLen=atoi(argv[c+1]);
				if(outLen<2 || outLen>pref->precision)
				{
					printf("Invalid argument for -o!\n\nThe argument must be a number between 2 and %i.\nFor example: -o 8\n\nPlease use man calc for further information\n",pref->precision);
					delete pref;
					delete[] vars;
					return 0;
				}
				pref->outputLength=outLen;
				
				c++;
			}
			else {
				printf("Invalid argument for -o!\n\nThe argument must be a number between 2 and %i.\nFor example: -o 8\n\nPlease use man calc for further information\n",pref->precision);
				delete pref;
				delete[] vars;
				return 0;
			}
		}
		else if(strcmp(argv[c],"-c")==0 || strcmp(argv[c],"--complex")==0)
		{
			pref->complex=true;
		}
		else if(strcmp(argv[c],"-h")==0 || strcmp(argv[c],"--help")==0)
		{
			printf("Usage:\ncalc \"input\"\nFor example: calc \"2*(3+7/2)\"\n\nOptions:\n-a, --angle (deg, rad, gra)\n-m, --mode (base, std, script)\n-c, --complex\n-b, --base (dec, hex, bin, oct) only in base mode\n-o, --output (2-18)\n-h, --help\n-v, --version\n\nPlease use man calc for  further information\n");
			delete pref;
			delete[] vars;
			return 0;		
		}
		else if(strcmp(argv[c],"-v")==0 || strcmp(argv[c],"--version")==0)
		{
			printf("\n%s\n\n%s",VERSIONSTRING,AUTHORSTRING);
			delete pref;
			delete[] vars;
			return 0;		
		}
		else {
			ask=false;
			if(input==NULL)
			{
				input=new char[strlen(argv[c])+1];
				memcpy(input,argv[c],strlen(argv[c])+1);
			}
			else {
				char*temp=input;
				input=new char[strlen(argv[c])+strlen(temp)+1];
				memcpy(input,temp,strlen(temp));
				memcpy(&input[strlen(temp)],argv[c],strlen(argv[c])+1);
			}
		}
	}
	if(pref->calcType == SCIENTIFIC && pref->base!=DEC)
	{
		printf("Invalid option -b!\nOption -b can only be used when -m is set!\n\nPlease use man calc for further information\n");
		delete pref;
		delete[] vars;
		return 0;
	}
	if(ask)
	{
		input=new char[1026];
		scanf("%1024s",input);
	}
	if(script)
	{
		char*path=input;
		struct stat fileinfo;
		FILE*scriptFile;
		if(lstat(path,&fileinfo)!=0)
		{
			printf("Opening script file %s failed!\n",path);
			delete[]path;
			return 0;
		}
		if(fileinfo.st_size<=0)
		{
			printf("Invalid file %s !\n",path);
			delete[]path;
			return 0;
		}
		scriptFile=fopen(path,"r");
		if(scriptFile==NULL)
		{
			printf("Opening script file %s failed!\n",path);
			delete[]path;
			return 0;
		}
		input = new char[fileinfo.st_size+1];
		if(fread(input,fileinfo.st_size,1,scriptFile)<=0)
		{
			printf("Can't read file %s !\n",path);
			delete[]path;
			return 0;
		}		
		input[fileinfo.st_size]=(char)0;
	}

	char*printString=new char[50];
	char*res=NULL;
		
	scriptData.error=false;
	
	struct termios terminfo,oldTerminfo;
	tcgetattr(fileno(stdout),&terminfo);
	oldTerminfo=terminfo;
	terminfo.c_lflag &=~ECHO;
	terminfo.c_lflag &=~ICANON;
	if(tcsetattr(fileno(stdout),TCSANOW,&terminfo)!=0)
		perror("tcsetattr fehler");

	scriptData.status=0;
	scriptData.usleep=false;
	scriptData.bbreak=false;
	scriptData.bcontinue=false;
	
	scriptData.exit=false;
	
	if(script)
	{
		
		searchScripts(input,pref,vars,&scriptData);
		scriptData.countDifference=0;
		initDebugging(input,&scriptData);
		int pret=preferencesPreprocessor(input,pref);
		res=checkString(input,pref);
		
		if(pret!=0)
		{
			switch(pret)
			{
				case PPINVALIDCOMMAND:
					fprintf(stderr,"\nPreference Preprocessor: Invalid preprocessor command\n");
					break;
				case PPINVALIDARGUMENT:
					fprintf(stderr,"\nPreference Preprocessor: Invalid preference argument\n");
					break;
				case PPINVALIDPREF:
					fprintf(stderr,"\nPreference Preprocessor: Invalid preference\n");
					break;	
				case PPEMPTY:
					fprintf(stderr,"\nPreference Preprocessor: Result script file empty\n");
					break;
				default:
					fprintf(stderr,"\nPreference Preprocessor: Unknown error\n");
					break;
			}
			delete[]input;
			delete[]printString;
			delete[]vars;
			delete pref;
			if(tcsetattr(fileno(stdout),TCSANOW,&oldTerminfo)!=0)
				perror("tcsetattr fehler");
			return 0;
		}


		

		scriptData.calcMode=false;
		fprintf(stderr,"\nProcessing main file ...\n");
		if(res==NULL)
		{
			fprintf(stderr,"Preprocessor Error\n");
			delete[]input;
			delete[]printString;
			delete[]vars;
			delete pref;
			if(tcsetattr(fileno(stdout),TCSANOW,&oldTerminfo)!=0)
				perror("tcsetattr fehler");
			return 0;
		}
		s=new Script((Script*)NULL,res,pref,vars,&scriptData);
		loadSubScripts(&scriptData,pref,vars,s);
		if(scriptData.error)
		{
			fprintf(stderr,"Script processing failed\n");
			delete[]input;
			delete[]printString;
			delete[]vars;
			delete pref;
			delete s;
			if(tcsetattr(fileno(stdout),TCSANOW,&oldTerminfo)!=0)
				perror("tcsetattr fehler");
			return 0;
		}
		value=s->exec();
	}
	else 
	{
		res=checkString(input,pref);
		scriptData.calcMode=true;
		s=new Script((Script*)NULL,res,pref,vars,&scriptData);
		value=s->exec();
	}
	

	
	if(pref->calcType==SCIENTIFIC)
	{	
		switch(value.type)
		{
			case NFLOAT:
			{
				if(value.fval.imag()==0.0)
				{
					printf("%.*Lg",pref->outputLength,real(value.fval));
				}
				else {
					if(value.fval.imag()>0.0)
						printf("%.*Lg+%.*Lgi",pref->outputLength,real(value.fval),pref->outputLength,imag(value.fval));
					else printf("%.*Lg%.*Lgi",pref->outputLength,real(value.fval),pref->outputLength,imag(value.fval));
				}
	
				break;
			}
			case NINT:
			{
				printf("%lli",value.ival);
				break;
			}
			case NBOOL:
				if(value.bval)
					printf("true");
				else printf("false");
				break;
			case NCHAR:
			{
				printf(value.cval);
				break;
			}
			default:
				printf("none");
				break;
		}
		printf("\n");
		//printf("%.*Lg\n",pref->outputLength,vars[26][0]=;
	}
	else {
		long long ival;
		
		switch(value.type)
		{
			case NFLOAT:
				ival=(long long)value.fval.real();
				break;
			case NINT:
				ival=value.ival;
				break;
			case NBOOL:
				ival=(long long)value.bval;
				break;
			default:
				ival=0;
		}
		
		
		switch(pref->base)
		{
			case DEC:
				printf("%lli\n",ival);
				break;
			case BIN:
			{
				long long result=ival;
				char*cRes=new char[65];
				cRes[64]=(char)0;
				for(long long c=0; c<64;c++)
				{
					unsigned long long match=1;
					match=match<<c;
					unsigned long long bin=match&result;
					if(bin==0)
						cRes[63-c]='0';
					else cRes[63-c]='1';
				}
				int c=0;
				while(cRes[c]=='0' && c<63)
					c++;
				printf("%s\n",&cRes[c]);
				break;
				}
			case HEX:
				printf("%LX\n",ival);
				break;
			case OCT:
				printf("%llo\n",ival);
				break;
		}	

	}

	char*newVarString="";

	char*varString=new char[maxLength+20];

	scriptData.vars[26][0]=value;

	for(int c=0; c<27; c++)
	{
		if(scriptData.vars[c][0].type==NINT)
			scriptData.vars[c][0].fval=Complex((long double)scriptData.vars[c][0].ival,0.0);
		else if(scriptData.vars[c][0].type==NBOOL)
			scriptData.vars[c][0].fval=Complex((long double)scriptData.vars[c][0].bval,0.0);
		else if(scriptData.vars[c][0].type!=NFLOAT)
			scriptData.vars[c][0].fval=Complex(NAN,0.0);
		
		sprintf(varString,"%.*Lg",maxLength,scriptData.vars[c][0].fval.real());
		newVarString=strins(newVarString,varString,strlen(newVarString));
		newVarString=strins(newVarString,"\n",strlen(newVarString));
		
		sprintf(varString,"%.*Lg",maxLength,scriptData.vars[c][0].fval.imag());
		newVarString=strins(newVarString,varString,strlen(newVarString));
		newVarString=strins(newVarString,"\n",strlen(newVarString));
	}
	chdir(getenv("HOME"));
	FILE*variablesFile=fopen(".calcvariables","w");
	if(variablesFile!=NULL)
	{
		fwrite(newVarString,strlen(newVarString),1,variablesFile);
		fclose(variablesFile);
	}

	if(tcsetattr(fileno(stdout),TCSANOW,&oldTerminfo)!=0)
		perror("tcsetattr fehler");
	
	delete[]input;
	delete[]res;
	delete[] scriptData.vars;		

	delete pref;
	delete[] vars;
	return 0;
}





void initDebugging(char*code,ThreadSync*scriptData)
{

	int pos1=0,pos2=0;
	bool quote=false;
	char* line;
	int len=strlen(code);
	
	int lineLen;
	int lineCount=0;
	int run=1;
	int c;
	while(run)
	{
		for(c=pos1;code[c]!='\n' && c<len+1; c++);
		pos2=c;
		if(c>=len)
			run=false;
		lineCount++;
		line=new char[pos2-pos1+1];
		memcpy(line,&code[pos1],pos2-pos1);
		line[pos2-pos1]=(char)0;
		lineLen=strlen(line);
		for(int c=0; c<lineLen && !(line[c]=='/' && line[c+1]=='/'); c++)
		{
			if(line[c]=='\"')
				quote=!quote;
			else if(!quote && line[c]==';')
			{
				scriptData->semicolonLines.NewItem(lineCount);
			}
		}
		pos1=pos2+1;
		delete[]line;
	}
}


void searchScripts(char*code,Preferences*pref,Variable*vars,ThreadSync*scriptData)
{
	char*cleanString=checkString(code,pref);
	if(cleanString==NULL)
		return;
	char*scriptName;
	int len=strlen(cleanString);
	int pos=0,pos2;
	FILE*subFile;
	while(pos<len)
	{
		if(strncmp(&cleanString[pos],"run(\"",4) == 0) //run("")
		{
			pos2=bracketFind(cleanString,")",pos+4);
			if(pos2-pos<7 || cleanString[pos2-1]!='\"')
			{
				pos++;
				continue;
			}
			scriptName=new char[pos2-pos-5];
			strcopy(scriptName,&cleanString[pos+5],pos2-pos-6);
			bool newScript=true;
			for(int c=0; c<scriptData->subprogramPath.GetLen(); c++)
				if(strcmp(scriptData->subprogramPath[c],scriptName)==0)
				{
					newScript=false;
					break;
				}
			if(newScript)
			{
				struct stat fileStat;
				char*subFileContent;

				if(lstat(scriptName,&fileStat)!=0)
					;
				else {
					subFile=fopen(scriptName,"r");
					if(subFile==NULL || fileStat.st_size<=0)
						;
					else
					{
						subFileContent=new char[fileStat.st_size+1];
						subFileContent[fileStat.st_size]=(char)0;
						fread(subFileContent,fileStat.st_size,1,subFile);
						fclose(subFile);
						scriptData->subprogramPath.NewItem(scriptName);
						searchScripts(subFileContent,pref,vars,scriptData);
						delete[]subFileContent;
					}
				}
			}
			pos=pos2+1;
		}
		else pos++;
	}
	delete[] cleanString;
}


void loadSubScripts(ThreadSync*scriptData,Preferences*pref,Variable*vars,Script*parent)
{
	struct stat fileStat;
	char*subFileContent;
	char*cleanSubFileContent;
	char*filePath;
	Script*subScript;
	FILE*subFile;
	for(int c=0; c<scriptData->subprogramPath.GetLen(); c++)
	{
		filePath=scriptData->subprogramPath[c];
		if(lstat(filePath,&fileStat)!=0)
			scriptData->subprograms.NewItem((Math*)NULL);
		else {
			subFile=fopen(filePath,"r");
			if(subFile==NULL || fileStat.st_size<=0)
				scriptData->subprograms.NewItem((Math*)NULL);
			else
			{
				subFileContent=new char[fileStat.st_size+1];
				subFileContent[fileStat.st_size]=(char)0;
				fread(subFileContent,fileStat.st_size,1,subFile);
				fclose(subFile);
				scriptData->countDifference+=scriptData->semicolonLines.GetLen();
				initDebugging(subFileContent,scriptData);
				cleanSubFileContent=checkString(subFileContent,pref);
				fprintf(stderr,"\nProcessing file ");
				fprintf(stderr,scriptData->subprogramPath[c]);
				fprintf(stderr,"\n");
				if(cleanSubFileContent==NULL)
				{
					fprintf(stderr,"Preprocessor Error\n");
					subScript=NULL;
				}
				else 
				{
					subScript=new Script(parent,NULL,pref,vars,scriptData);
					subScript->split(cleanSubFileContent);
				}
				delete[]subFileContent;
				scriptData->subprograms.NewItem(subScript);
			}
		}
	}

}


int preferencesPreprocessor(char *code,Preferences*pref)
{
	int pos=0,startPos=0,endPos=0,len;
	bool comment=false,quote=false;
	char*configString=NULL;
	
	
	while(pos<=(len=strlen(code)))
	{
		if(quote)
		{
			if(code[pos]=='\"')
				quote=false;
		}
		else if(comment)
		{
			if(code[pos]=='\n')
				comment=false;
		}
		else if(code[pos]=='/' && code[pos+1]=='/')
			comment=true;
		else if(code[pos]=='\"')
			quote=true;
		else if(code[pos]=='#')
		{
			startPos=pos;
			int configStartPos=startPos;
			endPos=pos;
			
			if(strncmp(&code[startPos],"#config",7)==0)
				configStartPos+=7;
			else return PPINVALIDCOMMAND;
			
			while(code[endPos]!='\n' && endPos<=len) endPos++;
			

			
			while((code[configStartPos]==' ' || code[configStartPos]=='\t') && configStartPos<=len) configStartPos++;
			
			configString=(char*)realloc(configString,endPos-configStartPos+1);
			strcopy(configString,&code[configStartPos],endPos-configStartPos);
		

			if(strncmp(configString,"complexon",9)==0)
				pref->complex=true;
			else if(strncmp(configString,"complexoff",10)==0)
				pref->complex=false;
			else if(strncmp(configString,"angledeg",8)==0)
				pref->angle=DEG;
			else if(strncmp(configString,"anglerad",8)==0)
				pref->angle=RAD;
			else if(strncmp(configString,"anglegra",8)==0)
				pref->angle=GRA;
			else if(strncmp(configString,"modebase",8)==0)
				pref->calcType=BASE;
			else if(strncmp(configString,"modescientific",14)==0)
				pref->calcType=SCIENTIFIC;
			else if(strncmp(configString,"clearmemory",11)==0)
				;//not used in calc
			else if(strncmp(configString,"outputlength",12)==0)
			{				int num=atoi(configString+12);
				if(num>=2 && num<=pref->precision)
					pref->outputLength=num;
				else return PPINVALIDARGUMENT;
			}
			else return PPINVALIDPREF;
			
			memmove(&code[startPos],&code[endPos],len-endPos+1);
			
			if(endPos==len)
				break;
			pos=startPos;
		}
		pos++;
	}
	
	if(strlen(code)<=0)
		return PPEMPTY;
	
	return 0;
}





